File: luogu/P1197-dsu.py
def get_parent(nodes, node):
    if nodes[node] == node:
        return node
    else:
        par = nodes[node]
        root = get_parent(nodes, par)
        nodes[node] = root
        return root

def merge(nodes, nA, nB):
    parA, parB = get_parent(nodes, nA), get_parent(nodes, nB)
    if parA == parB:
        return False
    else:
        nodes[parB] = parA
        return True

n, m = input().split()
n, m = int(n), int(m)

edges = [[] for _ in range(n)]
for i in range(m):
    x, y = input().split()
    x, y = int(x), int(y)
    edges[x].append(y)
    edges[y].append(x)

k = int(input())
cuihuis = []
for i in range(k):
    cuihuis.append(int(input()))

cuihuis_set = set(cuihuis)

nodes = [i for i in range(n)]
cnt_lian = n
for node in range(n):
    if node not in cuihuis_set:
        for nodeN in edges[node]:
            if nodeN not in cuihuis_set:
                if merge(nodes, node, nodeN):
                    cnt_lian -= 1

cnt_lian -= len(cuihuis_set)
results = [cnt_lian]
for c in cuihuis[::-1]:
    cuihuis_set.remove(c)
    cnt_lian += 1
    for node in edges[c]:
        if node not in cuihuis_set:
            if merge(nodes, c, node):
                cnt_lian -= 1
    results.append(cnt_lian)

for cnt in results[::-1]:
    print(cnt)

File: luogu/P1197-dsu.cpp
#include <iostream>
#include <vector>
using namespace std;
const int NMAX = 400010;
vector<int> edges[NMAX], cuihuis;
bool huifu[NMAX];
int unionset[NMAX];
int get_parent(int node) {
    if (unionset[node] == node) {
        return node;
    } else {
        int par = unionset[node];
        int root = get_parent(par);
        unionset[node] = root;
        return root;
    }
}
bool merge(int nA, int nB) {
    int parA = get_parent(nA), parB = get_parent(nB);
    if (parA != parB) {
        unionset[parB] = parA;
        return true;
    } else {
        return false;
    }
}
int main() {
    int n, m;
    scanf("%d%d", &n, &m);
    int x, y;
    for (int i = 0; i < m; i ++) {
        scanf("%d%d", &x, &y);
        edges[x].push_back(y);
        edges[y].push_back(x);
    }
    int k = 0;
    scanf("%d", &k);
    for (int i = 0; i < n; i++) {
        huifu[i] = true;
    }
    int c;
    for (int i = 0; i < k; i++) {
        scanf("%d", &c);
        cuihuis.push_back(c);
        huifu[c] = false;
    }

    for (int i = 0; i < n; i++) {
        unionset[i] = i;
    }
    int cnt_lian = n;
    for (int node = 0; node < n; node++) {
        if (huifu[node]) {
            for (int i = 0; i < edges[node].size(); i++) {
                int nodeN = edges[node][i];
                if (huifu[nodeN]) {
                    if (merge(node, nodeN)) {
                        cnt_lian --;
                    }
                }
            }
        }
    }

    cnt_lian -= k;
    vector<int> results;
    results.push_back(cnt_lian);
    for (int i = k-1; i >= 0; i--) {
        int node = cuihuis[i];
        huifu[node] = true;

        cnt_lian ++;
        for (int j = 0; j < edges[node].size(); j++) {
            int nodeN = edges[node][j];
            if (huifu[nodeN]) {
                if (merge(node, nodeN)) {
                    cnt_lian --;
                }
            }
        }

        results.push_back(cnt_lian);
    }

    for (int i = results.size()-1; i >= 0; i--){
        printf("%d\n", results[i]);
    }

    return 0;
}